icu4j/main/core/src/main/java/com/ibm/icu/impl/units/UnitsConverter.java

```diff
@@ -139,6 +196,54 @@ public BigDecimal convertInverse(BigDecimal inputValue) {
         return result;
     }
 
+    private static final BigDecimal[] minMetersPerSecForBeaufort = {
+        // Minimum m/s (base) values for each Bft value, plus an extra artificial value;
+        // when converting from Bft to m/s, the middle of the range will be used
+        // (Values from table in Wikipedia, except for artificial value).
+        // Since this is 0 based, max Beaufort value is thus array dimension minus 2.
+        BigDecimal.valueOf(0.0), // 0 Bft
+        BigDecimal.valueOf(0.3), // 1
+        BigDecimal.valueOf(1.6), // 2
+        BigDecimal.valueOf(3.4), // 3
+        BigDecimal.valueOf(5.5), // 4
+        BigDecimal.valueOf(8.0), // 5
+        BigDecimal.valueOf(10.8), // 6
+        BigDecimal.valueOf(13.9), // 7
+        BigDecimal.valueOf(17.2), // 8
+        BigDecimal.valueOf(20.8), // 9
+        BigDecimal.valueOf(24.5), // 10
+        BigDecimal.valueOf(28.5), // 11
+        BigDecimal.valueOf(32.7), // 12
+        BigDecimal.valueOf(36.9), // 13
+        BigDecimal.valueOf(41.4), // 14
+        BigDecimal.valueOf(46.1), // 15
+        BigDecimal.valueOf(51.1), // 16
+        BigDecimal.valueOf(55.8), // 17
+        BigDecimal.valueOf(61.4), // artificial end of range 17 to give reasonable midpoint
+    };
+
+    private BigDecimal scaleToBase(BigDecimal scaleValue, BigDecimal[] minBaseForScaleValues) {
+        BigDecimal pointFive = BigDecimal.valueOf(0.5);
+        BigDecimal scaleAdjust = scaleValue.abs().add(pointFive); // adjust up for later truncation
+        BigDecimal scaleAdjustCapped = scaleAdjust.min(BigDecimal.valueOf(minBaseForScaleValues.length - 2));
+        int scaleIndex = scaleAdjustCapped.intValue();
+        // Return midpont of range (the final range uses an articial end to produce reasonable midpoint)
+        return minBaseForScaleValues[scaleIndex].add(minBaseForScaleValues[scaleIndex + 1]).multiply(pointFive);
+    }
+
+    private BigDecimal baseToScale(BigDecimal baseValue, BigDecimal[] minBaseForScaleValues) {
+        int scaleIndex = Arrays.binarySearch(minBaseForScaleValues, baseValue.abs());
+        if (scaleIndex < 0) {
+            // since our first array entry is 0, this value will always be -2 or less
+            scaleIndex = -scaleIndex - 2;
+        }
```

Comment:
The return value of binarySearch is defined as "index of the search key, if it is contained in the array; otherwise, (-(insertion point) - 1). The insertion point is defined as the point at which the key would be inserted into the array: the index of the first element greater than the key, or a.length if all elements in the array are less than the specified key."

So since the range for Beaufort 0 is 0.0 to just under 0.3 m/s: If we were searching for 0.0 m/s. binarySearch would return 0, which we can use as the index directly. If we were searching for 0.1, the insertionPoint would be 1 (index of the first element greater than the key), so binarySearch would return -2, which we also want to map to an index of 0.

Reply:
That's a really poorly designed API, but that's obviously not your fault.  Thanks for the explanation!